File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/activity/commission_controller.py
import logging
from typing import Any
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from django.db import transaction
from ....configs.messages_templates import get_exception_class_message
from .._helpers.http import get_redirect_url
from ....Models.activity.commission import Commission
from ....Models.individuals.employee import Employee
from .._traits.controller import Controller
logger = logging.getLogger(__name__)
class CommissionController(Controller):
    
    def commission_create(self, request: HttpRequest, id: int) -> HttpResponse:
        logger.info("CommissionController.commission_create() called for employee ID: %s", id)
        try:
            employee: Employee = get_object_or_404(Employee, pk=id)
            commissions = Commission.commissiontype
            logger.debug("Commission types retrieved: %s", commissions)
            return render(request, 'commission/create.html', {
                'employee': employee,
                'commissions': commissions
            })
        except Exception as e:
            logger.exception("Error in commission_create: %s", e)
            messages.error(request, "An error occurred while loading the commission creation page.")
            return redirect(get_redirect_url(request))

    def store(self, request: HttpRequest) -> HttpResponse:
        logger.info("CommissionController.store() called by user: %s", request.user)
        try:
            if not request.user.has_perm('create commission'):
                logger.warning("User [%s] lacks 'create commission' permission.", request.user)
                messages.error(request, get_exception_class_message(PermissionDenied, __class__.__name__))
                return redirect(get_redirect_url(request))

            employee_id: Any = request.POST.get('employee_id')
            title: Any = request.POST.get('title')
            amount: Any = request.POST.get('amount')
            if not employee_id or not title or not amount:
                logger.error("Missing required field(s): employee_id=[%s], title=[%s], amount=[%s]", employee_id, title, amount)
                messages.error(request, "Employee, Title and Amount are required.")
                return redirect(get_redirect_url(request))

            # Critical financial operation: wrap in an atomic transaction.
            with transaction.atomic():
                commission: Commission = Commission()
                commission.employee_id = employee_id
                commission.title = title
                commission.type = request.POST.get('type')
                commission.amount = amount
                commission.created_by = request.user.creator_id()
                commission.save()
                logger.info("Commission created successfully with ID: %s", commission.id)
                
            messages.success(request, "Commission successfully created.")
            return redirect(get_redirect_url(request))
        except Exception as e:
            logger.exception("Error in store(): %s", e)
            messages.error(request, "An error occurred while creating commission.")
            return redirect(get_redirect_url(request))

    def show(self, request: HttpRequest, commission_id: int) -> HttpResponse:
        logger.info("CommissionController.show() called for commission ID: %s", commission_id)
        try:
            # The show method here redirects; you might add additional logging or processing.
            return redirect('commission_index')
        except Exception as e:
            logger.exception("Error in show(): %s", e)
            messages.error(request, "An error occurred while retrieving commission details.")
            return redirect(get_redirect_url(request))

    def edit(self, request: HttpRequest, commission_id: int) -> HttpResponse:
        logger.info("CommissionController.edit() called for commission ID: %s", commission_id)
        try:
            commission: Commission = get_object_or_404(Commission, pk=commission_id)
            if request.user.has_perm('edit commission'):
                if commission.created_by == request.user.creator_id():
                    commissions = Commission.commissiontype
                    logger.debug("Rendering edit view for commission ID: %s", commission_id)
                    return render(request, 'commission/edit.html', {
                        'commission': commission,
                        'commissions': commissions
                    })
                else:
                    logger.warning("User [%s] is not allowed to edit commission ID: %s", request.user, commission_id)
                    return JsonResponse({'error': "Permission denied."}, status=401)
            else:
                logger.warning("User [%s] lacks 'edit commission' permission.", request.user)
                return JsonResponse({'error': "Permission denied."}, status=401)
        except Exception as e:
            logger.exception("Error in edit(): %s", e)
            return JsonResponse({'error': "An error occurred."}, status=500)

    def update(self, request: HttpRequest, commission_id: int) -> HttpResponse:
        logger.info("CommissionController.update() called for commission ID: %s", commission_id)
        try:
            if not request.user.has_perm('edit commission'):
                logger.warning("User [%s] lacks 'edit commission' permission.", request.user)
                messages.error(request, get_exception_class_message(PermissionDenied, __class__.__name__))
                return redirect(get_redirect_url(request))

            commission: Commission = get_object_or_404(Commission, pk=commission_id)
            if commission.created_by != request.user.creator_id():
                logger.warning("User [%s] not authorized to update commission ID: %s", request.user, commission_id)
                messages.error(request, get_exception_class_message(PermissionDenied, __class__.__name__))
                return redirect(get_redirect_url(request))

            title: Any = request.POST.get('title')
            amount: Any = request.POST.get('amount')
            if not title or not amount:
                logger.error("Missing required update fields for commission ID [%s]: title=[%s], amount=[%s]", commission_id, title, amount)
                messages.error(request, "Title and Amount are required.")
                return redirect(get_redirect_url(request))

            # Critical update: using transaction.atomic() for ACID compliance.
            with transaction.atomic():
                commission.title = title
                commission.type = request.POST.get('type')
                commission.amount = amount
                commission.save()
                logger.info("Commission ID [%s] updated successfully.", commission_id)

            messages.success(request, "Commission successfully updated.")
            return redirect(get_redirect_url(request))
        except Exception as e:
            logger.exception("Error in update(): %s", e)
            messages.error(request, "An error occurred while updating commission.")
            return redirect(get_redirect_url(request))

    def destroy(self, request: HttpRequest, commission_id: int) -> HttpResponse:
        logger.info("CommissionController.destroy() called for commission ID: %s", commission_id)
        try:
            if not request.user.has_perm('delete commission'):
                logger.warning("User [%s] lacks 'delete commission' permission.", request.user)
                messages.error(request, get_exception_class_message(PermissionDenied, __class__.__name__))
                return redirect(get_redirect_url(request))

            commission: Commission = get_object_or_404(Commission, pk=commission_id)
            if commission.created_by != request.user.creator_id():
                logger.warning("User [%s] is not allowed to delete commission ID: %s", request.user, commission_id)
                messages.error(request, get_exception_class_message(PermissionDenied, __class__.__name__))
                return redirect(get_redirect_url(request))

            with transaction.atomic():
                commission.delete()
                logger.info("Commission ID [%s] deleted successfully.", commission_id)
            messages.success(request, "Commission successfully deleted.")
            return redirect(get_redirect_url(request))
        except Exception as e:
            logger.exception("Error in destroy(): %s", e)
            messages.error(request, "An error occurred while deleting commission.")
            return redirect(get_redirect_url(request))

File: _inc/laravel/app/Http/Controllers/Activity/CommissionController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
  DatabaseConstants,
  MiddlewaresConstants,
  UsersConstants,
  ViewsConstants
};
use App\Models\{Commission, Employee};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Contracts\View\View;
use Illuminate\Http\{Request, RedirectResponse};
use Illuminate\Support\Facades\{Auth, DB, Log, Route, View as ViewFacade};

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\HasCrudConstants;
use App\Traits\DefinesResourceActions;
class CommissionController extends Controller
{
	use DefinesResourceActions;

    use HasCrudConstants;

  use ChecksLogin;
  use ChecksPermissions;

  public function __construct()
  {
    $this->middleware(MiddlewaresConstants::AUTH);
  }

  public const COM_CR = 'commissionCreate';
  public function commissionCreate(Request $request, int|string $employeeId): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::COM . '.create';
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $employeeId, $action, $class, $viewPath) {
      try {
        Log::info("[{$class}::{$action}] start", [UsersConstants::COL_USER_ID => $req->user()->id, UsersConstants::COL_EMP_ID => $employeeId]);
        if ($resp = $this->_authorize($req, 'create commission')) {
          Log::warning("[{$class}::{$action}] unauthorized", [UsersConstants::COL_USER_ID => $req->user()->id]);
          return $resp;
        }
        $employee = Employee::find($employeeId);
        if (!$employee) {
          Log::warning("[{$class}::{$action}] employee not found", [UsersConstants::COL_EMP_ID => $employeeId]);
          return redirect()->back()->with('error', __('Employee not found.'));
        }
        $types = Commission::$commissionType;
        Log::info("[{$class}::{$action}] ready", [UsersConstants::COL_EMP_ID => $employee->id, 'types_count' => count($types)]);
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath, compact('employee', 'types'));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'employee_id' => $employeeId]);
  }

  public function index(Request $request): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::COM . '.index';
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      try {
        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        $user = $userOrRedirect;
        Log::info("[{$class}::{$action}] start", [UsersConstants::COL_USER_ID => $user?->id]);
        if ($resp = $this->_authorize($req, 'view commission')) {
          Log::warning("[{$class}::{$action}] unauthorized");
          return $resp;
        }
        $creatorId = $user?->creatorId();
        $commissions = Commission::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)->orderByDesc('id')->get();
        Log::info("[{$class}::{$action}] fetched commissions", ['count' => $commissions->count()]);
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath, compact('commissions'));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function create(Request $request, int|string|null $employeeId = null): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::COM . '.create';
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $employeeId, $action, $class, $viewPath) {
      try {
        Log::info("[{$class}::{$action}] start", [UsersConstants::COL_USER_ID => $req->user()->id, UsersConstants::COL_EMP_ID => $employeeId]);
        if ($resp = $this->_authorize($req, 'create commission')) {
          Log::warning("[{$class}::{$action}] unauthorized");
          return $resp;
        }
        $employee = Employee::find($employeeId);
        if (!$employee) {
          Log::warning("[{$class}::{$action}] employee not found", [UsersConstants::COL_EMP_ID => $employeeId]);
          return redirect()->back()->with('error', __('Employee not found.'));
        }
        $types = Commission::$commissionType;
        Log::info("[{$class}::{$action}] data ready", ['types' => count($types)]);
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath, compact('employee', 'types'));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'employee_id' => $employeeId]);
  }

  public function store(Request $request): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      try {
        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        $user = $userOrRedirect;
        Log::info("[{$class}::{$action}] start", [UsersConstants::COL_USER_ID => $user?->id, 'input' => $req->only([UsersConstants::COL_EMP_ID, 'title', 'type', 'amount'])]);
        if ($resp = $this->_authorize($req, 'create commission')) {
          Log::warning("[{$class}::{$action}] unauthorized");
          return $resp;
        }
        $valStart = microtime(true);
        $data = $req->validate([UsersConstants::COL_EMP_ID => 'required|exists:employees,id', 'title' => 'required|string', 'type' => 'required', 'amount' => 'required|numeric']);
        $this->logExecutionTime($valStart, $action, 'validateStore');
        $creatorId = $user?->creatorId();
        $txnStart = microtime(true);
        DB::transaction(function () use ($data, $creatorId, $action, $class) {
          $commission = Commission::create([UsersConstants::COL_EMP_ID => $data[UsersConstants::COL_EMP_ID], 'title' => $data['title'], 'type' => $data['type'], 'amount' => $data['amount'], DatabaseConstants::COL_TABLE_CREATOR => $creatorId]);
          Log::info("[{$class}::{$action}] created", ['commission_id' => $commission->id]);
        });
        $this->logExecutionTime($txnStart, $action, 'storeTransaction');
        return redirect()->back()->with('success', __('Commission successfully created.'));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function show(Commission $commission): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($commission, $action, $class) {
      Log::info("[{$class}::{$action}] redirecting", ['commission_id' => $commission->id]);
      return redirect()->route(ViewsConstants::COM . '.index');
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'commission_id' => $commission->id]);
  }

  public function edit(Request $request, int|string $id): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::COM . '.edit';
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $id, $action, $class, $viewPath) {
      try {
        Log::info("[{$class}::{$action}] start", ['id' => $id]);
        if ($resp = $this->_authorize($req, 'edit commission')) {
          Log::warning("[{$class}::{$action}] unauthorized");
          return $resp;
        }
        $commission = Commission::findOrFail($id);
        if ($resp = $this->authorizeOwnership($req, $commission)) return $resp;
        $types = Commission::$commissionType;
        Log::info("[{$class}::{$action}] data ready", ['commission_id' => $commission->id, 'types' => count($types)]);
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath, compact('commission', 'types'));
      } catch (AuthorizationException $e) {
        Log::error("[{$class}::{$action}] auth error", ['error' => $e->getMessage()]);
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'commission_id' => $id]);
  }

  public function update(Request $request, Commission $commission): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $commission, $action, $class) {
      try {
        Log::info("[{$class}::{$action}] start", ['commission_id' => $commission->id, 'input' => $req->only(['title', 'type', 'amount'])]);
        if ($resp = $this->_authorize($req, 'edit commission')) {
          Log::warning("[{$class}::{$action}] unauthorized");
          return $resp;
        }
        if ($resp = $this->authorizeOwnership($req, $commission)) return $resp;
        $valStart = microtime(true);
        $data = $req->validate(['title' => 'required|string', 'type' => 'required', 'amount' => 'required|numeric']);
        $this->logExecutionTime($valStart, $action, 'validateUpdate');
        $txnStart = microtime(true);
        DB::transaction(function () use ($commission, $data, $action, $class) {
          $lock = Commission::where('id', $commission->id)->lockForUpdate()->firstOrFail();
          $lock->update($data);
          Log::info("[{$class}::{$action}] updated", ['commission_id' => $commission->id]);
        });
        $this->logExecutionTime($txnStart, $action, 'updateTransaction');
        return redirect()->back()->with('success', __('Commission successfully updated.'));
      } catch (AuthorizationException $e) {
        Log::error("[{$class}::{$action}] auth error", ['error' => $e->getMessage()]);
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'commission_id' => $commission->id]);
  }

  public function destroy(Request $request, Commission $commission): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $commission, $action, $class) {
      try {
        Log::info("[{$class}::{$action}] start", ['commission_id' => $commission->id]);
        if ($resp = $this->_authorize($req, 'delete commission')) {
          Log::warning("[{$class}::{$action}] unauthorized");
          return $resp;
        }
        if ($resp = $this->authorizeOwnership($req, $commission)) return $resp;
        $txnStart = microtime(true);
        DB::transaction(function () use ($commission, $action, $class) {
          $lock = Commission::where('id', $commission->id)->lockForUpdate()->firstOrFail();
          $lock->delete();
          Log::info("[{$class}::{$action}] deleted", ['commission_id' => $commission->id]);
        });
        $this->logExecutionTime($txnStart, $action, 'destroyTransaction');
        return redirect()->back()->with('success', __('Commission successfully deleted.'));
      } catch (AuthorizationException $e) {
        Log::error("[{$class}::{$action}] auth error", ['error' => $e->getMessage()]);
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        Log::error("[{$class}::{$action}] unexpected error", ['error' => $e->getMessage()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'commission_id' => $commission->id]);
  }

  protected function authorizeOwnership(Request $request, Commission $commission): RedirectResponse|null
  {
    Log::info(__CLASS__ . '::' . __FUNCTION__ . ' start', [
      UsersConstants::COL_USER_ID       => $request->user()->id,
      'commission_id' => $commission->id
    ]);
    if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) {
      Log::warning(__CLASS__ . '::' . __FUNCTION__ . ' not logged in');
      return $userOrRedirect;
    }
    $user = $userOrRedirect;
    if ($commission->created_by !== $user?->creatorId()) {
      Log::warning(__CLASS__ . '::' . __FUNCTION__ . ' forbidden', [
        UsersConstants::COL_USER_ID          => $user?->id,
        'commission_owner' => $commission->created_by
      ]);
      return defaultPermissionDenial(
        $request,
        new AuthorizationException(),
        __CLASS__ . '::' . __FUNCTION__
      );
    }
    Log::info(__CLASS__ . '::' . __FUNCTION__ . ' ownership verified');
    return null;
  }

  /**
   * Authorize the current request for a given permission.
   *
   * Returns a redirect response when denied, or null when authorized.
   */
  private function _authorize(Request $request, string $ability): ?RedirectResponse
  {
    $result = self::guard($request, $ability);
    return $result === true ? null : $result;
  }
}
